### PR TITLE
Refactor next/head to React hooks

### PR DESCRIPTION
**Pull request by tannerlinsley**
Migrates the functionality that `side-effect` provided to the head component into a React hook

### CODE REVIEW

**Comment by tannerlinsley** [comment]
Clearly my TS-fu is really young here. So any guidance in that arena is much appreciated.

**Comment by tannerlinsley** [comment]
Using a lot of `any` here to get things to pass. If anyone has a few minutes to walk through the types with me, I would really appreciate it. I'll continue to work on them in the meantime.

File: packages/next-server/lib/head.tsx
Comment (timneutkens) [inline packages/next-server/lib/head.tsx]: It's actually better to write the check directly into the `if` statement because `typeof window` gets replaced at build time and tree shaking will be applied 💯

Comment (tannerlinsley) [reply packages/next-server/lib/head.tsx]: Totally. This was mostly a transfer from the side-effect file. But I agree 100%

Comment (timneutkens) [reply packages/next-server/lib/head.tsx]: That typeof replace is not standard (we recently added it) hence why I wanted to mention it in case you hadn't seen 💯 🚀

Comment (tannerlinsley) [reply packages/next-server/lib/head.tsx]: 👍

File: packages/next-server/lib/head.tsx
Comment (timneutkens) [inline packages/next-server/lib/head.tsx]: This is not the exact equivalent of https://github.com/zeit/next.js/pull/7937/files#diff-d2ba5b59eb8d9e5d8a2ea25c45833eceL50 right 🤔 So in this case it would trigger more than it would before?

Comment (tannerlinsley) [reply packages/next-server/lib/head.tsx]: Since there is no dependency array, this does fire after every render, exactly like `componentDidUpdate`.

Comment (Timer) [reply packages/next-server/lib/head.tsx]: I also second that this isn't exactly identical -- previously, `componentDidUpdate` would only fire `emitUpdate()`.

The current implementation, following the semantics of `useEffect`, means every component update will cause a cleanup and re-register, meaning an update now invokes:
```js
// old cleanup
headInstanceRefs.delete(instanceRef) // L175
emitUpdate() // L176
// new run
headInstanceRefs.add(instanceRef) // L164
emitUpdate() // L172
```

To get the exact behavior as previously, we'd almost need to split this into two effects:
```js
  // every component update, including mount, triggers emitUpdate()
  React.useEffect(() => {
    emitUpdate()
  })
  // register an on-unmount effect, n.b. `[]` dep array and a noop main function body
  React.useEffect(() => {
     return () => {
      headInstanceRefs.delete(instanceRef)
      emitUpdate()
    }
  }, [])
```

Comment (tannerlinsley) [reply packages/next-server/lib/head.tsx]: Ahhh. Yep, that makes more sense. I've pushed those changes. 👍

File: packages/next-server/lib/head.tsx
Comment (Timer) [inline packages/next-server/lib/head.tsx]: This needs to be in a on-mount-only effect.

```suggestion
  React.useEffect(() => {
    headInstanceRefs.add(instanceRef)
  }, [])
```

So it should probably be merged with the effect below and then re-ordered.

**Review by Timer — DISMISSED** [review]
Looks good to me, I'll let @timneutkens have one final review and we can merge this! Thanks so much for this work @tannerlinsley!

File: packages/next-server/lib/head.tsx
Comment (alexandernanberg) [inline packages/next-server/lib/head.tsx]: Not 100% sure but I think that mutating the ref in render will be "unsafe" in concurrent mode

Comment (Timer) [reply packages/next-server/lib/head.tsx]: Very good catch, mutation must happen in useEffect.

Comment (tannerlinsley) [reply packages/next-server/lib/head.tsx]: While this is very true for operations that could get out of sync between render-attempts and actual reconciler commits, eg. `ref.current++`, this shouldn't be an issue here as it's merely a synchronous way of keeping track of the latest `children` that have been passed to the component, and has no dependency on the previous value of the ref.

Comment (Timer) [reply packages/next-server/lib/head.tsx]: In concurrent mode a render can be thrown away and thus not reflective of the actual state. 🤔

Comment (tannerlinsley) [reply packages/next-server/lib/head.tsx]: If a previous render does get thrown away, we can still trust that the value will get updated again in a future render before the final commit. It's also valuable to look at how the ref is being used to know if it's safe or not to do this:
- `instanceRef` is being used synchronously in the render function during SSR below. While not technically possible yet, concurrent mode will likely become possible during SSR very soon, so we need to be mindful of that `typeof window === 'undefined'` code block. Luckily, the `instanceRef` we're using here is being added to a `Set` in that block (which will get deduped during multiple non-commit renders) and `emitUpdate` is immutably constructing the new head list regardless of the previous call to `emitUpdate`. With that said, it would still be safe that multiple renders happens before an actual commit on the server, even with a hypothetical concurrent mode.
- For the code blocks using `instanceRef` on the client, both blocks are wrapped in `useEffect`, so we're also safe that the value is the most up to date version of the `instanceRef` and that it will only get executed during the commit phase.

I can see that although this code is safe, that may not be very clear to most working on it, so we can talk about alternatives. One simple alternative could be adding the `instanceRef` update logic to each `useEffect` block that relies on it and also the `typeof window === 'undefined'` block that uses it in SSR. eg.

```diff
function Head({ children }: { children: JSX.Element }) {
  const ampState = React.useContext(AmpStateContext)
  const updateHead = React.useContext(HeadManagerContext)
  const instanceRef = React.useRef(children)

- // Update the instanceRef every render
- instanceRef.current = children

  const inAmpMode = isInAmpMode(ampState)

  const emitUpdate = () => {
    headState = reduceHeadInstances([...headInstanceRefs], {
      inAmpMode,
    })
    if (updateHead) {
      updateHead(headState)
    }
  }

  // The `useEffect` hook is only called on the client.
  // This means we need to simulate the component being mounted and trust that
  // `rewind()` is invoked to clean up these entries.
  if (typeof window === 'undefined') {
+   // Update the instanceRef
+   instanceRef.current = children
    headInstanceRefs.add(instanceRef)
    emitUpdate()
  }

  // We need to register this head instance on mount and unregister it when
  // unmounted.
  React.useEffect(() => {
+   // Update the instanceRef
+   instanceRef.current = children
    headInstanceRefs.add(instanceRef)
    // n.b. emitUpdate() for mount is called in below effect
    return () => {
      headInstanceRefs.delete(instanceRef)
      emitUpdate()
    }
  }, [])

  // Trigger an update on mount and every update (note the undefined dependency
  // array)
  React.useEffect(() => {
    emitUpdate()
  })

  return null
}
```

File: packages/next-server/lib/head.tsx
Comment (Jessidhia) [inline packages/next-server/lib/head.tsx]: This would be better done by making `reduceHeadInstances` accept `Iterable` and passing it through `Array.from` before reducing.

File: packages/next-server/lib/head.tsx
Comment (Jessidhia) [inline packages/next-server/lib/head.tsx]: This may generate duplicate `key`s as `React.Children.toArray` / `React.Children.map` [add `key`s to elements](https://github.com/facebook/react/blob/v16.9.0/packages/react/src/ReactChildren.js#L295-L306).

This also makes the `i` in `const key = c.key || i` unreachable.

File: packages/next-server/lib/head.tsx
Comment (Jessidhia) [reply packages/next-server/lib/head.tsx]: This suggestion incorrectly puts the `instanceRef.current` update in the second `Effect` (which only adds the object to the `Set`) instead of the third `Effect`, which is what actually would cause the `Set` to be read and needs to be updated on every render; but putting the assignments in the server-side block and the `Effect` block that calls `emitUpdate()` look sufficient.

The complication here is that different `<Head/>` components might commit at different times (depending on `Suspense` boundaries), but they all call `emitUpdate`; if multiple `Head` instances participate in a render pass, but one of them is inside a boundary that suspends, the `emitUpdate`s called by the `Head`s inside the subtree that did commit would see the `render`-updated `instanceRef.current`.

There probably is room here for optimizing updates to only handle each `Head`'s own `children`; the tricky part is SSR.

**Comment by timneutkens** [comment]
I'm going to close this PR in favor of #8981 as we'll have to deprecate `next/head` very soon. Thank you very much for creating this PR @tannerlinsley 🙏

**Pull request closed without merging**
